### PR TITLE
[Merged by Bors] - Add Holesky

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

Add the Holesky network config as per https://github.com/eth-clients/holesky/tree/36e4ff2d5138dcb2eb614f0f60fdb060b2adc1e2/custom_config_data.

Since the genesis state is ~190MB, I've opted to *not* include it in the binary and instead download it at runtime (see #4564 for context). To download this file we have:

- A hard-coded URL for a SigP-hosted S3 bucket with the Holesky genesis state. Assuming this download works correctly, users will be none the wiser that the state wasn't included in the binary (apart from some additional logs)
- If the user provides a `--checkpoint-sync-url` flag, then LH will download the genesis state from that server rather than our S3 bucket.
- If the user provides a `--genesis-state-url` flag, then LH will download the genesis state from that server regardless of the S3 bucket or `--checkpoint-sync-url` flag.
- Whenever a genesis state is downloaded it is checked against a checksum baked into the binary.
- A genesis state will never be downloaded if it's already included in the binary.
- There is a `--genesis-state-url-timeout` flag to tweak the timeout for downloading the genesis state file.

## Log Output

Example of log output when a state is downloaded:

```bash
Aug 23 05:40:13.424 INFO Logging to file                         path: "/Users/paul/.lighthouse/holesky/beacon/logs/beacon.log"
Aug 23 05:40:13.425 INFO Lighthouse started                      version: Lighthouse/v4.3.0-bd9931f+
Aug 23 05:40:13.425 INFO Configured for network                  name: holesky
Aug 23 05:40:13.426 INFO Data directory initialised              datadir: /Users/paul/.lighthouse/holesky
Aug 23 05:40:13.427 INFO Deposit contract                        address: 0x4242424242424242424242424242424242424242, deploy_block: 0
Aug 23 05:40:13.427 INFO Downloading genesis state               info: this may take some time on testnets with large validator counts, timeout: 60s, server: https://sigp-public-genesis-states.s3.ap-southeast-2.amazonaws.com/
Aug 23 05:40:29.895 INFO Starting from known genesis state       service: beacon
```

Example of log output when there are no URLs specified:

```
Aug 23 06:29:51.645 INFO Logging to file                         path: "/Users/paul/.lighthouse/goerli/beacon/logs/beacon.log"
Aug 23 06:29:51.646 INFO Lighthouse started                      version: Lighthouse/v4.3.0-666a39c+
Aug 23 06:29:51.646 INFO Configured for network                  name: goerli
Aug 23 06:29:51.647 INFO Data directory initialised              datadir: /Users/paul/.lighthouse/goerli
Aug 23 06:29:51.647 INFO Deposit contract                        address: 0xff50ed3d0ec03ac01d4c79aad74928bff48a7b2b, deploy_block: 4367322
The genesis state is not present in the binary and there are no known download URLs. Please use --checkpoint-sync-url or --genesis-state-url.
```

## Additional Info

I tested the `--genesis-state-url` flag with all 9 Goerli checkpoint sync servers on https://eth-clients.github.io/checkpoint-sync-endpoints/ and they all worked 🎉 

My IDE eagerly formatted some `Cargo.toml`. I've disabled it but I don't see the value in spending time reverting the changes that are already there.

I also added the `GenesisStateBytes` enum to avoid an unnecessary clone on the genesis state bytes baked into the binary. This is not a huge deal on Mainnet, but will become more relevant when testing with big genesis states.

When we do a fresh checkpoint sync we're downloading the genesis state to check the `genesis_validators_root` against the finalised state we receive. This is not *entirely* pointless, since we verify the checksum when we download the genesis state so we are actually guaranteeing that the finalised state is on the same network. There might be a smarter/less-download-y way to go about this, but I've run out of cycles to figure that out. Perhaps we can grab it in the next release?

### CODE REVIEW

**Comment by realbigsean** [comment]
FYI there was discussion on the CL call about serving this state snappy compressed in checkpointz servers

File: common/eth2_network_config/built_in_network_configs/holesky/boot_enr.yaml
Comment (antondlr) [inline common/eth2_network_config/built_in_network_configs/holesky/boot_enr.yaml:1]: additional bootnode ENR: 
`- enr:-Le4QCRIcJJgCJiK8fOwIUQp6eNz5D-rbj7R4WHtglRJRYT2eGc5fqwtPeW67Op56DbADEV8ojLe9PbdLgrQ9oUad9gChGV0aDKQqX6DZjABcAAKAAAAAAAAAIJpZIJ2NIJpcIQ5gR6Wg2lwNpAgAUHQBwEQAAAAAAAAADR-iXNlY3AyNTZrMaEDb_TxT3XRqQ4rX_9OFB1kvuwscxe8ztsHqyjkkPOBbmeDdWRwgiMohHVkcDaCI4I`

**Comment by paulhauner** [comment]
> FYI there was discussion on the CL call about serving this state snappy compressed in checkpointz servers

Thanks, I'm currently lobbying against that: https://discord.com/channels/595666850260713488/745077610685661265/1144388115230887996

File: common/eth2_network_config/built_in_network_configs/holesky/config.yaml
Comment (jimmygchen) [inline common/eth2_network_config/built_in_network_configs/holesky/config.yaml]: Updated to 256 in https://github.com/eth-clients/holesky/pull/59

```suggestion
CAPELLA_FORK_EPOCH: 256
```

File: common/eth2_network_config/built_in_network_configs/holesky/boot_enr.yaml
Comment (jimmygchen) [reply common/eth2_network_config/built_in_network_configs/holesky/boot_enr.yaml:1]: Is this a Lighthouse bootnode? might be worth adding it here too https://github.com/eth-clients/holesky/blob/main/custom_config_data/boot_enr.txt

File: beacon_node/src/config.rs
Comment (jimmygchen) [inline beacon_node/src/config.rs]: Just a note here - It would be possible for user to only specify "checkpoint-sync-url" and "checkpoint-sync-url-timeout" and get a timeout if checkpoint sync server doesn't respond with the genesis state within `genesis_state_url_timeout`. Not sure if we should change anything though.

File: common/eth2_config/src/lib.rs
Comment (jimmygchen) [inline common/eth2_config/src/lib.rs]:
```suggestion
    /// The genesis state for this network is not yet known.
```

File: common/eth2_network_config/built_in_network_configs/holesky/boot_enr.yaml
Comment (paulhauner) [reply common/eth2_network_config/built_in_network_configs/holesky/boot_enr.yaml:1]: Done: https://github.com/eth-clients/holesky/pull/62

Comment (paulhauner) [reply common/eth2_network_config/built_in_network_configs/holesky/boot_enr.yaml:1]: I've also added the ENR to this PR: c62f8b159

Comment (paulhauner) [reply common/eth2_network_config/built_in_network_configs/holesky/boot_enr.yaml:1]: Eh, we need to regenerate the boot ENR because the Capella fork epoch was changed.

File: common/eth2_config/src/lib.rs
Comment (jimmygchen) [inline common/eth2_config/src/lib.rs]:
```suggestion
    ($(($name_ident: ident, $config_dir: tt, $genesis_state_source: path)),+) => {
        $(
        define_archive!($name_ident, $config_dir, $genesis_state_source);
```

File: beacon_node/src/config.rs
Comment (paulhauner) [reply beacon_node/src/config.rs]: Good point! I've fixed that in 78f073165

**Comment by paulhauner** [comment]
FYI I just bumped the default `--genesis-state-url-timeout` to match the existing `--checkpoint-sync-url-timeout` (180s). I don't really see why they should be different.

See 766a1c360

**Comment by paulhauner** [comment]
I've labelled this PR as `blocked` since we need to regenerate the our ENR to suit the Capella epoch change (https://github.com/sigp/lighthouse/pull/4653/commits/1a0fccf7791eda40c671b1c64060d5aac9ebf1ff).

This branch will now generate the new ENR. I'm not exactly sure if we need to delete the old `network` directory to regenerate the key.. that might be safest.

File: common/eth2_network_config/src/lib.rs
Comment (jimmygchen) [inline common/eth2_network_config/src/lib.rs:338]: It could be a nice feature to support `GenesisStateSource::Url` for custom testnets in the future - but may be an overkill since users have to download some other files anyway.

File: lighthouse/src/main.rs
Comment (jimmygchen) [inline lighthouse/src/main.rs]: Perhaps worth mentioning that checkpoint-sync-url will be used if this is not supplied?

File: beacon_node/src/config.rs
Comment (jimmygchen) [inline beacon_node/src/config.rs:486]: Does `genesis_state_url` actually get used if `--checkpoint-sync-url` is supplied? Wouldn't it just go with checkpoint sync because we get `ClientGenesis::CheckpointSyncUrl` instead of `ClientGenesis::GenesisState` below?

https://github.com/sigp/lighthouse/blob/766a1c360d25698682d9f59e5c448ace300b7078/beacon_node/src/config.rs#L519-L526

File: common/eth2_network_config/src/lib.rs
Comment (paulhauner) [reply common/eth2_network_config/src/lib.rs:338]: Yeah it would be cool. We'd need to add a new file to define the URL, so I thought I'd leave that for future works.

File: lighthouse/src/main.rs
Comment (paulhauner) [reply lighthouse/src/main.rs]: Thanks, fixed in d18edb7c5

File: beacon_node/src/config.rs
Comment (paulhauner) [reply beacon_node/src/config.rs:486]: We still fetch the genesis state when doing a checkpoint sync: https://github.com/paulhauner/lighthouse/blob/d18edb7c51ec03e903dbff4148255bdddd6806e0/beacon_node/client/src/builder.rs#L380

I have a note in the PR description that talks about this too:

> When we do a fresh checkpoint sync we're downloading the genesis state to check the genesis_validators_root against the finalised state we receive. This is not entirely pointless, since we verify the checksum when we download the genesis state so we are actually guaranteeing that the finalised state is on the same network. There might be a smarter/less-download-y way to go about this, but I've run out of cycles to figure that out. Perhaps we can grab it in the next release?

Comment (jimmygchen) [reply beacon_node/src/config.rs:486]: Ah I see, I missed that. Thanks!

File: common/eth2_network_config/built_in_network_configs/holesky/boot_enr.yaml
Comment (paulhauner) [reply common/eth2_network_config/built_in_network_configs/holesky/boot_enr.yaml:1]: I've gone and regenerated an ENR on our Holesky boot node, so need for you to do it Anton ☺️

**Comment by paulhauner** [comment]
I've updated our ENR to be Capella-256-ready. I also added [Teku's ENRs](https://github.com/eth-clients/holesky/pull/63/files).

See https://github.com/sigp/lighthouse/pull/4653/commits/a944a7e97bd38f788026476e34df033a21608663

**Comment by paulhauner** [comment]
Alrighty, this one is now ready to go.

File: account_manager/src/validator/slashing_protection.rs
Comment (michaelsproul) [inline account_manager/src/validator/slashing_protection.rs]: It might be good (perhaps when removing Goerli's genesis) to embed the `genesis_validators_root` in the network config so we don't need to load it in cases like this. The GVR is often used for differentiating networks (including during signing), so it might be handy to have it.

File: common/eth2_network_config/built_in_network_configs/holesky/config.yaml
Comment (michaelsproul) [inline common/eth2_network_config/built_in_network_configs/holesky/config.yaml]:
```suggestion
# Sep-15-2023 13:55:00 +UTC
```

Update this assuming my PR upstream gets merged: https://github.com/eth-clients/holesky/pull/66

File: lighthouse/src/main.rs
Comment (michaelsproul) [inline lighthouse/src/main.rs:328]: There are no CLI tests for these new flags

**Review by michaelsproul — APPROVED** [review]
Looks good to me! Just a few small nitpicks, then I think we're ready to merge :rocket:

File: common/eth2_network_config/built_in_network_configs/holesky/boot_enr.yaml
Comment (jimmygchen) [inline common/eth2_network_config/built_in_network_configs/holesky/boot_enr.yaml]: It looks like these two EF ENRs don't exactly match the ones on the holesky repo, has something changed?:

https://github.com/eth-clients/holesky/blob/76057d57ab1f585519ecb606a9e5f7780e925a37/custom_config_data/boot_enr.txt#L2-L3

```suggestion
- enr:-KG4QF6d6vMSboSujAXTI4vYqArccm0eIlXfcxf2Lx_VE1q6IkQo_2D5LAO3ZSBVUs0w5rrVDmABJZuMzISe_pZundADhGV0aDKQqX6DZjABcAAAAQAAAAAAAIJpZIJ2NIJpcISygIjpiXNlY3AyNTZrMaEDF3aSa7QSCvdqLpANNd8GML4PLEZVg45fKQwMWhDZjd2DdGNwgiMog3VkcIIjKA
- enr:-Ly4QJLXSSAj3ggPBIcodvBU6IyfpU_yW7E9J-5syoJorBuvcYj_Fokcjr303bQoTdWXADf8po0ssh75Mr5wVGzZZsMBh2F0dG5ldHOIAAAAAAAAAACEZXRoMpCpfoNmMAFwAAABAAAAAAAAgmlkgnY0gmlwhJK-DYCJc2VjcDI1NmsxoQJrIlXIQDvQ6t9yDySqJYDXgZgLXzTvq8W7OI51jfmxJohzeW5jbmV0cwCDdGNwgiMog3VkcIIjKA
```

Comment (paulhauner) [reply common/eth2_network_config/built_in_network_configs/holesky/boot_enr.yaml]: Oh nice catch, that's probably the Capella fork update.

File: lighthouse/src/main.rs
Comment (paulhauner) [reply lighthouse/src/main.rs:328]: Good catch, I added some here: https://github.com/sigp/lighthouse/pull/4653/commits/9bdc2433320d02ce8f8db34fcd9dfbfddd3d7f0c

I wanted to test the behaviour with `--checkpoint-sync-url` however our test don't support it (they try to download a checkpoint state). I don't have the time to go down that rabbit hole at present.

File: account_manager/src/validator/slashing_protection.rs
Comment (paulhauner) [reply account_manager/src/validator/slashing_protection.rs]: Good call, I added that in [add8bd6](https://github.com/sigp/lighthouse/pull/4653/commits/add8bd6f43a08dddcd49a2e34656db2a127f2e45)

**Review by michaelsproul — APPROVED** [review]
Happy to merge once merge conflicts are resolved!

**Comment by paulhauner** [comment]
> Happy to merge once merge conflicts are resolved!

Thanks, I'll see how CI goes and then merge at will. Thanks for the review @jimmygchen and @michaelsproul 🙏

**Pull request closed without merging**
